Remove CFG_USERINPUTFD and use stdin replacement in affected tests.

antonsviridenko opened the pull request:
Closes #1609 and closes #1857

Affected tests now use stdin temporarily replaced with a pipe read end.

antonsviridenko committed:
Remove CFG_USERINPUTFD and use stdin replacement in affected tests.

Interactive key generation in rnpkeys now always reads answers from
stdin; tests feed it via a pipe (saving and restoring the original
stdin). Also sets CLOEXEC on the restored descriptor.

Closes #1609 and closes #1857

antonsviridenko commented:
```
 258/259 Test #248: rnp_tests.test_fuzz_verify ....................................................   Passed    5.99 sec
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
ci/lib/install_functions.inc.sh: line 676:  2438 Aborted                 (core dumped) "$@"
Error: Process completed with exit code 134.
```

antonsviridenko commented:
strange, all of the failing centos-and-fedora instances  fail with the same `std::bad_alloc` error in a place not possibly related to the introduced changes

ni4 commented:
@antonsviridenko It could be somehow related to the redirected stdin (and be thrown from GTest side), as other PRs work well.

antonsviridenko commented:
All failed instances use OpenSSL as a backend...

antonsviridenko commented:
Any ideas how to debug this?

```
ci/lib/install_functions.inc.sh: line 676:  2438 Aborted                 (core dumped) "$@"
```
Looks like it's something related to Python environment in Centos/Fedora.

ni4 commented:
@antonsviridenko Did you try to revert you commits, related to stdio redirection, and see whether that helps?

antonsviridenko commented:
> Did you try to revert you commits, related to stdio redirection, and see whether that helps?

I assume that yes, reverting commits will remove CI failures.

I think it could be related to duplicated STDIN not having CLOEXEC flags... And affected platforms run cli_tests through some python wrapper.

antonsviridenko commented:
Is there any possibility to recreate the same local environment as in failing centos-and-fedora machines?

ni4 commented:
@antonsviridenko Wow, this one get lost somehow. Locally you may just use corresponding Docker container, install prerequisites, copy rnp sources there, build and run tests. Another option is https://github.com/nektos/act but I didn't try it yet and don't know in which way it works and is it possible to properly debug things.

ronaldtse commented:
Rebased onto latest main (conflict in rnpcfg.h resolved, keeping main's crypto-refresh and allow-sha1 additions). The two commits were squashed into one with the following review fixes folded in:

- Removed a leftover debug fprintf(stderr, "saved_stdin: ...") in generatekey.cpp.
- Stdin is now always restored after rnp_cmd (the original only restored on the success path, leaving the test process's stdin redirected after failures).
- Fixed a C++ goto-over-initialization compile error in the new stdin handling.

Verified locally: full suite passes (304/304 rnp_tests + cli_tests).

ronaldtse commented:
Just need one more approval!

ronaldtse committed:
tests: fix generatekey.cpp build on Windows

The fcntl()/F_SETFD/FD_CLOEXEC call in ask_expert_details() is not
available on Windows (MSVC and MinGW), breaking the rnp_tests build.
Guard it with #ifndef _WIN32, matching the existing precedent in
ffi.cpp.

ronaldtse committed:
style: fix clang-format alignment in rnpcfg.h

Removing CFG_USERINPUTFD changed the longest entry of the CFG_*
define block, so clang-format now requires the trailing comments of
the whole block to be re-aligned.

ronaldtse commented:
Rebased onto current `main` (3e5f675b) — clean rebase, no conflicts. The previous run's typos failure (`priamry`) and the macOS `cli_tests-Encryption` failure were both old-base issues already fixed on main.

Two fix commits added on top (original commit kept intact):
- `72bbcb19` **tests: fix generatekey.cpp build on Windows** — the new stdin-restore code in `ask_expert_details()` uses `fcntl()`/`F_SETFD`/`FD_CLOEXEC`, which do not exist on Windows (MSVC and MinGW); this broke the `rnp_tests` build on all windows-msys2 and windows-native legs. Guarded with `#ifndef _WIN32`, matching the existing precedent in `ffi.cpp`.
- `cf4da353` **style: fix clang-format alignment in rnpcfg.h** — removing `CFG_USERINPUTFD` changed the longest entry of the `CFG_*` define block, so clang-format-11 requires the whole block's trailing comments to be re-aligned (applied verbatim from the CI clang-format output).

Verified locally: full suite passes 282/282 (rnp_tests + all cli_tests), Botan 3 build. The windows-msys2 legs are now green.

**fuzzing job: pre-existing repo-wide failure, not caused by this PR.** It fails at CMake configure: `Could NOT find Botan: Found unsuitable version "3.4.0", but required is at least "3.6.0"` — the oss-fuzz build image still ships Botan 3.4.0 while main requires >= 3.6.0 when crypto-refresh is enabled (`src/lib/CMakeLists.txt:35`). The identical failure occurs on other recent branches (sig-error-strings, fix-cli-log-redaction, ci-fix-main) because google/oss-fuzz#15882 ("[rnp] Update Botan to 3.6.0") is still open — the revert to `google/oss-fuzz@master` (3e5f675b) landed before the upstream PR was merged. Suggest restoring the rnpgp oss-fuzz fork workaround on main until google/oss-fuzz#15882 is merged.

ronaldtse committed:
tests: clearerr(stdin) in ask_expert_details()

On glibc, once stdin's EOF flag is set, subsequent fgets() calls fail
without reading, even after fd 0 is replaced with a fresh pipe. An
expected-failure expert keygen run (e.g. an invalid curve choice on
brainpool-less builds, as in RHEL/Fedora CI containers) reads past
the provided input and sets the flag, so the next ask_expert_details()
call failed immediately: rnpkeys_generatekey_testExpertMode broke on
the centos-and-fedora CI legs. Reset the flags after redirecting
stdin to the input pipe.

ronaldtse commented:
Update on the one remaining red leg: `windows-msys2` (ucrt64, botan) — `cli_tests-SignDefault.test_gpg_to_rnp_default_key` fails at `gpg_verify_cleartext` with `gpg: BAD signature from "signing@gpg"`.

Analysis: this is a pre-existing, content-random msys2/GnuPG flake, not caused by this PR:

1. The failing step is **gpg verifying gpg's own cleartext signature** — rnp is not involved in that step at all (`rnp_cleartext_signing_gpg_to_rnp`: gpg --clearsign → rnp --verify (passed) → gpg --verify (BAD)). No stdin is passed to any process on this path (the passphrase is a command-line argument).
2. The signed content is random per run (`random_text()` includes ` `, `\t`, `\r`, `\n`; sizes up to 1 MB), so the failure manifests nondeterministically: the first CI round failed `test_rnp_to_gpg_default_key` (rnp-signed, openssl leg), this round `test_gpg_to_rnp_default_key` (gpg-signed, botan leg), and the middle round was green on **all** msys2 legs including this one.
3. The only delta between that green round and this round is the `clearerr(stdin)` commit, which touches `src/tests/generatekey.cpp` — compiled only into `rnp_tests.exe`, not into the `rnp.exe`/`rnpkeys.exe` binaries driven by cli_tests.py — so it cannot have caused this failure.
4. The same suite flaked the same way on #2296's msys2 leg (failed once, passed on rerun).
5. Not reproducible off-Windows: 500 iterations of gpg --clearsign → gpg --verify over `random_text()`-style content (POSIX and Windows-text-mode-simulated `\r\r\n` content) with GnuPG 2.5.20 → 0 failures. The msys2 leg runs GnuPG 2.4.9; the trigger appears specific to that platform (Windows CR/LF handling of the random content).

Re-running the failed leg. A proper follow-up on main (unrelated to this PR) would be to make `random_text()` deterministic or avoid `\r`/trailing-whitespace edge cases in the cleartext tests.

ronaldtse committed:
tests: harden stdin restoration in ask_expert_details

Two robustness polish items on top of the CFG_USERINPUTFD removal:

* Guard close(saved_stdin) with a saved_stdin != -1 check. If the initial
  dup() failed, saved_stdin is -1; the unconditional close() was technically
  POSIX-valid (EBADF) but inelegant.

* Expand the FD_CLOEXEC restoration comment to explain WHY it is needed:
  dup() drops close-on-exec, dup2() inherits it from source, so without
  restoration any child process spawned by later tests would inherit a
  non-close-on-exec stdin.

Comments + defensive guard; no behaviour change. All 23 generate-key
tests pass.

ronaldtse commented:
## On the `codecov/project` (-0.07%) failure

The 0.07% coverage drop is a side-effect of the test-infrastructure change in `src/tests/generatekey.cpp` (`ask_expert_details`), not a real regression in library code.

### Why coverage drops

Replacing `CFG_USERINPUTFD` with a `dup2(stdin)` dance introduced several `goto end` error paths in the test helper:

- `dup(STDIN_FILENO)` failure
- `dup2(...)` failure on the replace
- `dup2(...)` failure on the restore
- `fcntl(STDIN_FILENO, F_SETFD, FD_CLOEXEC)` failure (POSIX only)

Each of these error paths is genuine defensive code (matches the existing POSIX-dance idioms elsewhere in the test suite), but none can be exercised without fault injection — `dup`/`dup2`/`fcntl` don't fail on a healthy runner. Hence uncovered branches, hence the small coverage drop.

### Why we should accept it

- The PR's purpose is **test infrastructure cleanup** (removing the `CFG_USERINPUTFD` escape hatch that leaked test affordances into the production CLI's config vocabulary). Coverage on production library code is unchanged.
- The new error paths mirror the POSIX-dance patterns already accepted in sibling test helpers.
- The drop (-0.07%) is well below the typical "real regression" threshold; it's a coverage-line-budget issue, not a behavior issue.
- Mocking `dup`/`dup2`/`fcntl` to exercise the error paths would require either `LD_PRELOAD` shims or a significant test-harness refactor — both out of scope for this cleanup PR.

Recommend treating the codecov check as a known-accepted regression for this PR.

ni4 reviewed (approved):
LGTM!

ronaldtse commented:
Thank you @antonsviridenko for the contribution!

ronaldtse commented:
Merging with 2 approvals (ni4, ronaldtse). Thank you @antonsviridenko for the contribution, and @ni4 for the review!
